x-pack/plugins/monitoring/server/plugin.ts

```diff
@@ -0,0 +1,324 @@
+import { i18n } from '@kbn/i18n';
+import { has, get } from 'lodash';
+import { UsageCollectionSetup } from 'src/plugins/usage_collection/server';
+import { TelemetryCollectionManager } from 'src/legacy/core_plugins/telemetry/server/collection_manager';
+import { LOGGING_TAG, KIBANA_MONITORING_LOGGING_TAG } from '../common/constants';
+import {
+  Logger,
+  PluginInitializerContext,
+  RequestHandlerContext,
+  KibanaRequest,
+  KibanaResponseFactory,
+  CoreSetup,
+  ICustomClusterClient,
+  ElasticsearchConfig,
+  ElasticsearchConfigType,
+} from '../../../../src/core/server';
+import { MonitoringConfig } from './config';
+// @ts-ignore
+import { requireUIRoutes } from './routes';
+// @ts-ignore
+import { instantiateClient } from './es_client/instantiate_client';
+// @ts-ignore
+import { initBulkUploader, registerCollectors } from './kibana_monitoring';
+// @ts-ignore
+import { initInfraSource } from './lib/logs/init_infra_source';
+import { registerMonitoringCollection } from './telemetry_collection';
+import { XPackMainPlugin } from '../../../legacy/plugins/xpack_main/server/xpack_main';
+import { LicensingPluginSetup } from '../../licensing/server';
+import { PluginSetupContract } from '../../features/server';
+import { LicenseService } from './license_service';
+import { MonitoringLicenseService } from './types';
+
+export interface LegacyAPI {
+  xpackMain: XPackMainPlugin;
+  telemetryCollectionManager: TelemetryCollectionManager;
+  elasticsearch: any;
+  opsInterval: number;
+  getOSInfo: () => any;
+  hapiServer: any;
+  getServerStatus: () => string;
+  serverEvents: any;
+  infra: any;
+}
+
+interface PluginsSetup {
+  usageCollection: UsageCollectionSetup;
+  licensing: LicensingPluginSetup;
+  features: PluginSetupContract;
+}
+
+interface MonitoringCoreConfig {
+  get: (key: string) => string;
+}
+interface MonitoringCore {
+  config?: () => MonitoringCoreConfig;
+  log?: Logger;
+  route?: (options: any) => void;
+  xpackInfo?: any;
+}
+
+export class Plugin {
+  private readonly initializerContext: PluginInitializerContext;
+  private readonly log: Logger;
+  private readonly getLogger: (...scopes: string[]) => Logger;
+  private cluster?: ICustomClusterClient;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.initializerContext = initializerContext;
+    this.log = initializerContext.logger.get(LOGGING_TAG);
+    this.getLogger = (...scopes: string[]) => initializerContext.logger.get(LOGGING_TAG, ...scopes);
+  }
+
+  async setup(core: CoreSetup, plugins: PluginsSetup) {
+    const [config, legacyConfig] = await combineLatest([
+      this.initializerContext.config.create<MonitoringConfig>(),
+      this.initializerContext.config.legacy.globalConfig$,
+    ])
+      .pipe(first())
+      .toPromise();
+
+    // Monitoring creates and maintains a connection to a potentially
+    // separate ES cluster - create this first
+    const elasticsearchConfig = new ElasticsearchConfig(
```

Comment:
it's easy to forget to wrap raw config in `ElasticsearchConfig` class. why not do it on the plugin boundaries?
https://github.com/elastic/kibana/blob/56ba753db0090b7e1109fc14b35eecfaf08719a3/src/core/server/elasticsearch/elasticsearch_service.ts#L66-L68

Reply:
I honestly don't know how that works.

How do I modify my existing code to do that?

```
const config = await this.initializerContext.config
      .create<MonitoringConfig>()
      .pipe(first())
      .toPromise();
```

Reply:
I believe it's:
```js
await this.initializerContext.config
      .create<MonitoringConfig>()
      .pipe(
         first(),
         map(rawConfig => new ElasticsearchConfig(rawConfig))
       )
      .toPromise();
```

Reply:
I don't think that will work. The entire config isn't a `ElasticsearchConfig` - just two separate parts: `ui.elasticsearch.` and `elasticsearch.`. Is there a way to "cast" part of the config to a typed config?

Reply:
Well... Everything I can think of sounds like a hack.
I still not quite understand why we don't want to keep an `elasticsearch config`-compatible interface? Why do we need to handle an empty array and undefined differently?